File: sdk/rust/src/providers/emailnator.rs
/*!
 * Emailnator 渠道实现
 * 网站: https://www.emailnator.com
 * 需要 XSRF-TOKEN Session 认证
 */

use crate::config::{block_on, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use serde_json::Value;

const BASE_URL: &str = "https://www.emailnator.com";
const EMAIL_OPTIONS: &[&str] = &["plusGmail", "dotGmail", "googleMail"];

fn browser_ua() -> &'static str {
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36"
}

fn init_session() -> Result<(String, String), String> {
    block_on(async {
        let resp = http_client()
            .get(BASE_URL)
            .header(
                "Accept",
                "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            )
            .header("User-Agent", browser_ua())
            .send()
            .await
            .map_err(|e| format!("emailnator init session failed: {}", e))?;

        let mut xsrf_token = String::new();
        let mut cookie_parts: Vec<String> = Vec::new();

        for value in resp.headers().get_all("set-cookie") {
            let cookie_str = value.to_str().unwrap_or("");
            if let Some(name_value) = cookie_str.split(';').next() {
                cookie_parts.push(name_value.to_string());
                if name_value.starts_with("XSRF-TOKEN=") {
                    let raw = &name_value["XSRF-TOKEN=".len()..];
                    xsrf_token = urlencoding::decode(raw).unwrap_or_default().to_string();
                }
            }
        }

        if xsrf_token.is_empty() {
            return Err("Failed to extract XSRF-TOKEN".into());
        }

        Ok((xsrf_token, cookie_parts.join("; ")))
    })
}

fn post_json(xsrf_token: &str, cookies: &str, path: &str, body: Value) -> Result<String, String> {
    let xsrf_token = xsrf_token.to_string();
    let cookies = cookies.to_string();
    let path = path.to_string();
    block_on(async move {
        let resp = http_client()
            .post(format!("{}{}", BASE_URL, path))
            .header("User-Agent", browser_ua())
            .header("Content-Type", "application/json")
            .header("Accept", "application/json, text/plain, */*")
            .header("Origin", BASE_URL)
            .header("Referer", format!("{}/", BASE_URL))
            .header("X-Requested-With", "XMLHttpRequest")
            .header("X-XSRF-TOKEN", xsrf_token)
            .header("Cookie", cookies)
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("emailnator {} failed: {}", path, e))?;

        let status = resp.status();
        let text = resp
            .text()
            .await
            .map_err(|e| format!("emailnator {} read failed: {}", path, e))?;
        if !status.is_success() {
            return Err(format!("emailnator {} failed: {} {}", path, status, text));
        }
        Ok(text)
    })
}

pub fn generate_email() -> Result<EmailInfo, String> {
    let (xsrf_token, cookies) = init_session()?;
    let text = post_json(
        &xsrf_token,
        &cookies,
        "/generate-email",
        serde_json::json!({ "email": EMAIL_OPTIONS }),
    )?;
    let data: Value = serde_json::from_str(&text).map_err(|e| format!("parse failed: {}", e))?;
    let email_list = data["email"]
        .as_array()
        .ok_or("emailnator: empty email response")?;

    if email_list.is_empty() {
        return Err("emailnator: empty email list".into());
    }

    let email = email_list[0].as_str().unwrap_or("").to_string();
    let token_payload = serde_json::json!({
        "xsrfToken": xsrf_token,
        "cookies": cookies,
    })
    .to_string();

    Ok(EmailInfo {
        channel: Channel::Emailnator,
        email,
        token: Some(token_payload),
        expires_at: None,
        created_at: None,
    })
}

pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let session: Value =
        serde_json::from_str(token).map_err(|e| format!("parse session token failed: {}", e))?;

    let xsrf_token = session["xsrfToken"].as_str().unwrap_or("");
    let cookies = session["cookies"].as_str().unwrap_or("");

    let text = post_json(
        xsrf_token,
        cookies,
        "/message-list",
        serde_json::json!({ "email": email }),
    )?;
    let data: Value = serde_json::from_str(&text).map_err(|e| format!("parse failed: {}", e))?;
    let message_data = data["messageData"].as_array().cloned().unwrap_or_default();

    let mut result = Vec::new();
    for msg in &message_data {
        let msg_id = msg["messageID"].as_str().unwrap_or("");
        if msg_id.starts_with("ADS") {
            continue;
        }
        let html = if msg_id.is_empty() {
            String::new()
        } else {
            let detail = post_json(
                xsrf_token,
                cookies,
                "/message-list",
                serde_json::json!({ "email": email, "messageID": msg_id }),
            )
            .unwrap_or_default();
            serde_json::from_str::<String>(&detail).unwrap_or(detail)
        };
        let raw = serde_json::json!({
            "id": msg_id,
            "from": msg["from"].as_str().unwrap_or(""),
            "to": email,
            "subject": msg["subject"].as_str().unwrap_or(""),
            "text": "",
            "html": html,
            "date": msg["time"].as_str().unwrap_or(""),
            "isRead": false,
            "attachments": [],
        });
        result.push(normalize_email(&raw, email));
    }

    Ok(result)
}

File: sdk/php/src/Providers/Emailnator.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * Emailnator 渠道实现 — https://www.emailnator.com
 *
 * XSRF-TOKEN Session 认证：GET / 从 Set-Cookie 提取 XSRF-TOKEN 与 session cookie；
 *   POST /generate-email（JSON {email:[options]}）返回 {email:[addr]}；
 *   POST /message-list（JSON {email}）拉列表，带 messageID 再 POST 拉正文；
 *   token 存 JSON{xsrfToken, cookies}。过滤 messageID 以 "ADS" 开头的广告。
 */
final class Emailnator
{
    private const BASE_URL = 'https://www.emailnator.com';

    private const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/145.0.0.0 Safari/537.36';

    /** @var string[] */
    private const EMAIL_OPTIONS = ['plusGmail', 'dotGmail', 'googleMail'];

    /**
     * @return array<string,string>
     */
    private static function defaultHeaders(): array
    {
        return [
            'User-Agent' => self::UA,
            'Accept' => 'application/json',
            'Content-Type' => 'application/json',
            'X-Requested-With' => 'XMLHttpRequest',
        ];
    }

    /**
     * 初始化 Session，返回 [xsrfToken, cookieStr]。
     *
     * @return array{0:string,1:string}
     */
    private static function initSession(): array
    {
        $resp = HttpClient::get(self::BASE_URL, ['User-Agent' => self::UA]);

        $cookies = [];
        $xsrf = '';
        foreach ($resp->getHeader('Set-Cookie') as $sc) {
            $first = trim(explode(';', $sc, 2)[0]);
            if ($first === '' || !str_contains($first, '=')) {
                continue;
            }
            [$k, $v] = explode('=', $first, 2);
            $k = trim($k);
            $v = trim($v);
            if ($k === '') {
                continue;
            }
            $decoded = rawurldecode($v);
            $cookies[$k] = $decoded;
            if ($k === 'XSRF-TOKEN') {
                $xsrf = $decoded;
            }
        }
        if ($xsrf === '') {
            throw new \RuntimeException('emailnator: 无法提取 XSRF-TOKEN');
        }
        $parts = [];
        foreach ($cookies as $k => $v) {
            $parts[] = $k . '=' . $v;
        }
        return [$xsrf, implode('; ', $parts)];
    }

    public static function generate(): EmailInfo
    {
        [$xsrf, $cookies] = self::initSession();

        $resp = HttpClient::post(
            self::BASE_URL . '/generate-email',
            self::defaultHeaders() + ['X-XSRF-TOKEN' => $xsrf, 'Cookie' => $cookies],
            json: ['email' => self::EMAIL_OPTIONS],
        );
        $data = HttpClient::json($resp);
        $emailList = $data['email'] ?? null;
        if (!is_array($emailList) || empty($emailList)) {
            throw new \RuntimeException('emailnator: 生成邮箱失败，响应为空');
        }

        $tokenPayload = (string) json_encode(['xsrfToken' => $xsrf, 'cookies' => $cookies], JSON_UNESCAPED_SLASHES);
        return new EmailInfo('emailnator', (string) $emailList[0], $tokenPayload);
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token): array
    {
        $session = json_decode((string) $token, true);
        if (!is_array($session)) {
            throw new \InvalidArgumentException('emailnator: token 格式无效');
        }
        $xsrf = (string) ($session['xsrfToken'] ?? '');
        $cookies = (string) ($session['cookies'] ?? '');

        $resp = HttpClient::post(
            self::BASE_URL . '/message-list',
            self::defaultHeaders() + ['X-XSRF-TOKEN' => $xsrf, 'Cookie' => $cookies],
            json: ['email' => $email],
        );
        $data = HttpClient::json($resp);
        $messageData = $data['messageData'] ?? null;
        if (!is_array($messageData)) {
            return [];
        }

        $out = [];
        foreach ($messageData as $msg) {
            if (!is_array($msg)) {
                continue;
            }
            $msgId = (string) ($msg['messageID'] ?? '');
            if (str_starts_with($msgId, 'ADS')) {
                continue;
            }
            $html = $msgId !== '' ? self::fetchDetail($xsrf, $cookies, $email, $msgId) : '';
            $raw = [
                'id' => $msgId,
                'from' => $msg['from'] ?? '',
                'to' => $email,
                'subject' => $msg['subject'] ?? '',
                'text' => '',
                'html' => $html,
                'date' => $msg['time'] ?? '',
                'isRead' => false,
                'attachments' => [],
            ];
            $out[] = Normalize::email($raw, $email);
        }
        return $out;
    }

    /**
     * POST /message-list 带 messageID 拉取单封 HTML 正文，失败返回空串。
     */
    private static function fetchDetail(string $xsrf, string $cookies, string $email, string $msgId): string
    {
        try {
            $resp = HttpClient::post(
                self::BASE_URL . '/message-list',
                self::defaultHeaders() + ['X-XSRF-TOKEN' => $xsrf, 'Cookie' => $cookies],
                json: ['email' => $email, 'messageID' => $msgId],
            );
        } catch (\Throwable) {
            return '';
        }
        $text = (string) $resp->getBody();
        $decoded = json_decode($text, true);
        if (is_string($decoded)) {
            return $decoded;
        }
        if (is_array($decoded)) {
            return (string) json_encode($decoded, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
        }
        return $text;
    }
}
